### PR TITLE
Move checks to field declaration directly

### PR DESCRIPTION
**Pull request by serenity4**
Will need more tests, but it might be a good starting point.

### CODE REVIEW

File: src/CheckConcreteStructs.jl [hunk 1]
```diff
@@ -15,25 +15,45 @@ function get_struct_name(expr::Expr)
...
-            ts = Base.return_types(s -> getfield_stable(s, vi), ($(esc(S)),))
-            t = only(ts)
-            if !isconcretetype(t)
-                error("Field $n is not concrete ($t)")
-            end
+    block = expr.args[3]
+    for (i, arg) in enumerate(block.args)
+        Meta.isexpr(arg, :(::)) || isa(arg, Symbol) || continue
+        if isa(arg, Symbol)
+            block.args[i] = :(throw($TypeNotConcreteError($(QuoteNode(arg)), Any)))
```
Comment (gdalle) [inline src/CheckConcreteStructs.jl:25]: Since we know that `arg` is not a `Symbol`, is this dead code?

File: src/CheckConcreteStructs.jl [hunk 2]
```diff
@@ -15,25 +15,45 @@ function get_struct_name(expr::Expr)
...
+        else
+            name, type = (arg.args[1], arg.args[2])
+            block.args[i] = Expr(:(::), name, :($check_is_concretetype($type, $(QuoteNode(name)))))
```
Comment (gdalle) [inline src/CheckConcreteStructs.jl:28]: Why is `QuoteNode` needed? I never really understood what it does

File: src/CheckConcreteStructs.jl [hunk 3]
```diff
@@ -15,25 +15,45 @@ function get_struct_name(expr::Expr)
...
         end
     end
+    esc(expr)
```
Comment (gdalle) [inline src/CheckConcreteStructs.jl:31]: Why do you choose to modify `expr` in place? We could just perform the checks, throw an error if needed, and then evaluate `expr` as it was?

**Review by gdalle — COMMENTED** [review]
Thanks, this looks really good!

File: src/CheckConcreteStructs.jl (hunk 2 above)
Comment (gdalle) [inline src/CheckConcreteStructs.jl:28]: Why is QuoteNode needed?

File: src/CheckConcreteStructs.jl (hunk 3 above)
Comment (serenity4) [reply src/CheckConcreteStructs.jl:31]: It is to move the checks at runtime, so we don't need `eval`.

File: src/CheckConcreteStructs.jl (hunk 2 above)
Comment (serenity4) [reply src/CheckConcreteStructs.jl:28]: To allow for a more informative error at runtime (e.g. passing `the name of a field `:x` to `TypeNotConcreteError`)

File: src/CheckConcreteStructs.jl (hunk 3 above)
Comment (gdalle) [reply src/CheckConcreteStructs.jl:31]: Just to be sure I understand, the checks actually happen during the first compilation of the block of code containing the macro-ed struct, right?

**Pull request merged**